### ISSUE TITLE
2.5.59.8 no more custom M106 at vertical layer bar?

### ISSUE THREAD

**Issue by CuredPrusa** [supermerill/SuperSlicer#4225]
### What happened?

I see that slicer introduces automatically M106 or M107 at every loop. When infill_fan_speed is not -1, the code gets M106 at every G command (each line). This cancels any attempt to input M106 at certain layer height. Is this behaviour planned for newer versions?  

### Project file & How to reproduce

![image](https://github.com/supermerill/SuperSlicer/assets/98842248/13e2343d-a549-49bf-8d4a-1a9c5876912a)

### Version

2.5.59.8

### Operating system

W10

### Printer model

V400 / Klipper

**Comment by supermerill** [comment]
You can set a very small fan speedup time (printer -> general -> cooling fan), like 0.01. That way, this post-porcessing is activated and will remove any duplicate of M106.

Maybe I should always run it. I'll do that for next version.
I didn't yet, as I wasn't very confident of it messing with gcode. Now it's more stable and shouldn't create bad gcode anymore.

**Comment by CuredPrusa** [comment]
I have tried your latest nightly (nightly from two days ago).
It's still not working correctly. But I have tried your proposal. With speedup 0 it's obvious M106 gets overwritten instantly: 

![image](https://github.com/supermerill/SuperSlicer/assets/98842248/b4bd3693-dd1e-4821-b208-b036d099d078)

With speedup 0,01  there is no custom code value at all. :
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/d9baba69-2ee2-477e-bc77-38648004e899)

**Comment by CuredPrusa** [comment]
>did you spot an issue?
>your fan speed has been moved up, as the move was very short. 
If you refer to line 62538 in first picture, that was not a move as short as one line command. In that layer there should be hundreds of command lines (G1) to complete the layer. And yet, fan speed got instantly retrieved to default value (line 62540) 

>What is the issue?
in second picture you see the beginning of the same layer as first picture, but with speedup set to 0,01. In instruction line 62546 the custom code got announced but the value is nowhere to be found.

**Comment by Rewwan** [comment]
I confirm the user code M106 cannot be set, it is constantly overwritten.

**Comment by Alexxx775** [comment]
The M106 command is installed, but then overwritten and does not work properly as it should.

**Comment by supermerill** [comment]
Can't reproduce, maybe give me your test project?
![image](https://github.com/supermerill/SuperSlicer/assets/6536403/3e16ceaa-7ecd-4e89-a9e2-e41a8a1083a6)

**Comment by Rewwan** [comment]
![1](https://github.com/supermerill/SuperSlicer/assets/72386243/a289fd7d-fb64-4ee5-a04b-bb763f428158)
![2](https://github.com/supermerill/SuperSlicer/assets/72386243/cab7a315-b772-4d95-a644-472cc73fa8de)
[3mf.zip](https://github.com/supermerill/SuperSlicer/files/15388697/3mf.zip)
Here is a project, it has custom code M106 S255 installed on layer 65. In version 2.4.58.5, custom code M106 was applied to all subsequent ones, and in the latest version there is a permanent override.

**Comment by supermerill** [comment]
your gcode file you sent me was made in the working 2.4.58.5, right?
can't reproduce in my nightly (my nightly in the left, your gcode in the right):
![image](https://github.com/supermerill/SuperSlicer/assets/6536403/fccf7877-a0d3-4c21-9055-818ad0139926)
what's the problematic version?

**Comment by Rewwan** [comment]
![2 4 58 5](https://github.com/supermerill/SuperSlicer/assets/72386243/73b2faf5-62bb-4190-8f5e-cd93dae7e86d)
![2 5 59 10](https://github.com/supermerill/SuperSlicer/assets/72386243/e38351d4-62b2-42c0-9715-b9b5cba4f9ac)
[Code.zip](https://github.com/supermerill/SuperSlicer/files/15393732/Code.zip)
It was chopped up in the new version. Here are two files from different versions of the slicer, in them the M106 S255 behaves differently

**Comment by supermerill** [comment]
Seems like it's working as intended, now. no?

**Comment by Rewwan** [comment]
No, in 2.4 the code is applied to all subsequent layers, and in 2.5 the airflow is constantly changing, even if the values ​​in the settings for perimeters and padding are set to -1, the default airflow is applied instead of the custom code as before.

**Comment by supermerill** [comment]
understood

**Comment by supermerill** [comment]
behaviour (in next nightly):
* if an override is -1, it's considered that it is at the default fan speed.
* if the default speed is -1 (disabled) then:
 * there won't be a M107 start, unless 'disable fan for the first layer'
 * if a fan override is not -1, at this feature the fan will be set + M107 at the end of the feature.
 * for fan_below_layer_time, -1 is considered as 0

So now (if defautl fan is -1) if you put a `M106 S255`, the fan won't be erased until a feature that isn't disabled

Is it okay?

**Comment by Rewwan** [comment]
Yes, I just checked the nightly build.  The code is applied and remains.  Thanks!

**Comment by CuredPrusa** [comment]
No, I don't think this should be a final solution, it's more like a temporary workaround.
I don't understand ( and forgive my ignorance please) why does new version have to write M106 constantly (in normal fan mode), while in 2.4 the setpoint for fan speed was written only in the beginning and that's all. So fan starts ( 2.4) in the beginning of print (as it should be) and then you can do with it whatever you want afterwards. Now, with this new nightly you can manipulate fan only if you disabled the fan. So I have to start fan  manually in the beginning of print if I want to change it later on.

Moreover, I think you should consider every printer's ability to adjust the fan speed during printing. Constantly writing the fan speed by slicer is disabling this important feature, especially useful when following the printing process during prototyping.  If someone wants to run fan 50% all the time without too much thinking, but then in the middle of print he want's to increase it to 100% because it's melting, he can not do that from printer if using the latest version of SS. 

I don't know anything about coding so it's quite interesting to see that such simple sounding feature (which all other slicers have by default) could be so tricky: KEEP THE FAN RUNNING AT SETPOINT UNTIL NEW SETPOINT GETS INTRODUCED MANUALLY (M106 - from manual gcode line or from printer override). 

Yes, there are many additional fan speed tweaks in new SS which I appreciate a lot. That could be the root cause of all the hassle. I don't think new features should disable basic crucial workflow of settings. I'd rather live without them and keep the solid fan commands of 2.4.

**Comment by supermerill** [comment]
> Moreover, I think you should consider every printer's ability to adjust the fan speed during printing.

like the speed? then it's on the firmware side to just interpolate with your %, like for the speed.

> KEEP THE FAN RUNNING AT SETPOINT UNTIL NEW SETPOINT GETS INTRODUCED MANUALLY 

In the nightly, it's now smart enough to not write it, and it's smart enough to detect the M106 from custom gcode.
If you want just a fan command at the start and then change it manually, then just write the command in the custom gcode. It's simple enough (and disable all fan settings).

**Comment by CuredPrusa** [comment]
>like the speed? then it's on the firmware side to just interpolate with your %, like for the speed.

Control Panel on my V400 sends M106 over Klipper when I want to manually adjust fan speed from the panel.  And then instantly gets annulled by frequently generated M106 from g-code.

>In the nightly, it's now smart enough to not write it, and it's smart enough to detect the M106 from custom gcode.

Unfortunately this is still not true. In picture below is how nightly writes gcode for top fill which should have fan at S76.5 (30%). At  every single non extruded move (travel) it is writing default fan speed S229.5 (90%). Why? Why can't it write one M106 S76.5 at the beginning of top fill and then write default M106 when top fill ends? So, there is NO WAY  to interfere with fan speed by sending one M106 from the panel/firmware. Perhaps this is happening because this new version is smart not to write surplus M106 down the gcode lines, but it writes it as soon as it detects there was a M106 change. 

![image](https://github.com/supermerill/SuperSlicer/assets/98842248/c478a207-b57a-43fe-b0f2-f467387b0216)

And that affects the good old easy manual M106 input at vertical scroll bar which version 2.4 process so naturally. 
Here, new nightly just waits for a new M106 intruder to annul it asap. 

 
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/077583d0-f6d7-4e3d-8fb8-300bbcb8c004)

>If you want just a fan command at the start and then change it manually, then just write the command in the custom gcode. It's simple enough (and disable all fan settings).

If this is your final solution about the manual M106 input which worked perfectly in 2.4, than I guess I should stick to 2.4 or Prusa for projects that require fan speed manipulation at critical layers. I certainly don't intend to write fan speeds for 12 layers in beginning ( I have one project which uses gradual fan increase during 12 layers, otherwise it gets loose from the bed) if this can be automated as it always was.

**Comment by Nbeknel** [comment]
I've also noticed when using the arachne engine the slicer will insert M106 commands before and after every width change.
![изображение](https://github.com/supermerill/SuperSlicer/assets/57685377/49952b4e-7ed3-404c-9a69-1c1ae10b68e4)
I've been able to reduce the amount by using a g-code substitution: `M106 S\d*(\.\d*)?\n;` -> `;` (regex), but now all M106 commands call the same value, even when it's printing the same feature.
![изображение](https://github.com/supermerill/SuperSlicer/assets/57685377/809b0d37-73c3-41e6-a87b-58e16dde6ef3)

**Comment by gmakedonski** [comment]
We are printing concrete, filament was set to ABS and fans speeds are :
![image](https://github.com/supermerill/SuperSlicer/assets/5514324/49d52981-d997-4d3e-b862-e4fa7242b06f)

but after every width move 
![image](https://github.com/supermerill/SuperSlicer/assets/5514324/1602ac64-dbd5-4459-9aa5-081ca3718c73)

**Comment by gmakedonski** [comment]
For me not fixed again on every width : M107

![image](https://github.com/supermerill/SuperSlicer/assets/5514324/b76a9071-30c3-451b-bebe-ce8ed19c100b)

attached slicerconfig and 3mf project
[3DCP.zip](https://github.com/user-attachments/files/15597291/3DCP.zip)

**Comment by supermerill** [comment]
> For me not fixed again on every width : M107

sorry, it's only for the ninghtly, in the 60.0 version.

**Comment by CuredPrusa** [comment]
Here I am again. 
I'm sorry if this could be annoying, but it certainly is to me.
Here's a project opened in 2.5.59.12. Blue squares are pointing at M106 commands. There are 25.928 M106 commands and 635 M107 commands.

![image](https://github.com/supermerill/SuperSlicer/assets/98842248/d2034004-7099-42f6-87ed-4c15b2003dc3)
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/75eef5e5-30fe-42ef-a26f-491144f81365)
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/929277e2-68c7-40b9-928d-03be197e3fff)
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/04f573af-fec1-4b86-bd8c-a8001e9c6007)
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/cd75c683-8eb4-44af-b754-caaff55dc086)

The same project opened in 2.4.58.5 has 4 M106 commands and 7 M107 commands.
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/343a9ad2-77cb-49b6-893d-3731a5dd8a70)
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/7ec7f234-1a9d-4ff1-9dd5-323ce3a9854b)
![image](https://github.com/supermerill/SuperSlicer/assets/98842248/e7bb501b-8b97-40e0-b1aa-2c633786be41)

I'm not sure this doesn't influence the processing power during printing.

**Comment by CuredPrusa** [comment]
Should I check latest nightly?

**Comment by CuredPrusa** [comment]
I think I see what is going on here... As many new controls you have added to the fan cooling, that many times the command gets repeated in front of every possible fan speed change. If certain move has a possibility to get a new command for fan, but there's no change (-1), then fan command is still written but with default value. True?

**Comment by CuredPrusa** [comment]
A note, not to miss the momentum...
in latest version 2.5.59.12, when ALL fan settings are at -1, gcode is packed with M107 like a pomegranate.

**Comment by supermerill** [comment]
it's better in the nightly, I added some code to not repeat itself.
I'll let the stable version as-is as it's more secure.
